### PR TITLE
[WIP][Prototype] Add streaming support to YARPC (through GRPC)

### PR DESCRIPTION
**Pull request by willhug**
Summary: This PR represents a first pass at what a streaming API would
look like in yarpc.  I'm modeling the API mostly off of the GRPC stream
APIs.

This has a few sections:

- Adding "Stream" as an option in the API surface (I've marked any
breaking changes we'll need to work around).
- Updating the Dispatcher to support streaming.
- Adding a raw encoding Stream API surface.
- Adding GRPC support to expose the streaming APIs.
- Adding a "streaming" example client/server that can be used to test
  the streaming APIs.

Since this PR does have breaking API changes in a few areas (*cough*
ClientConfig *cough*) it is not landable yet, but there's a lot of
feedback that can be given.

Test Plan: I've been using the "streaming" example to ensure that the
client and server can interact with each other, but, as we refine this
we'll need stricter testing.

### CODE REVIEW

File: api/transport/stream.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+// BaseStream is an interface for interacting with a stream.
+// TODO Should the Send/Recv functions pass around bytes/interface{} instead of
+// io.Readers?
+type BaseStream interface {
+	// Context returns the context for the stream.
+	Context() context.Context
+
+	// Request contains all the metadata about the request (without the body)
+	// TODO Don't use the transport request here, we should have another piece
```
Comment (peter-edge) [inline api/transport/stream.go]: Let's get whatever you want to do finalized

File: api/transport/stream.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// ServerStream represents the Server API of interacting with a Stream.
+type ServerStream interface {
+	BaseStream
+}
+
+// ClientStream represents the Client API of interacting with a Stream.
+type ClientStream interface {
+	// Close implements the io.Closer interface.  It closes the stream
+	// connection from the client side.
+	Close() error
```
Comment (peter-edge) [inline api/transport/stream.go]: grpc-go calls this `CloseSend`, it would be nice to match that

File: api/transport/stream.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+	Context() context.Context
+
+	// Request contains all the metadata about the request (without the body)
+	// TODO Don't use the transport request here, we should have another piece
+	// of info here to hold the connection metadata (basically the headers)
+	Request() *Request
+
+	// SendMsg sends a request over the stream. It blocks until the message
+	// has been sent.
+	SendMsg(m io.Reader) error
```
Comment (peter-edge) [inline api/transport/stream.go]: Recap from offline (just posting this): I'd like to be able to include `BaseStream, ClientStream, ServerStream` in generated protobuf code, but I don't like how grpc-go (and this) has `SendMsg/RecvMsg` on the included API, as I don't want end-users using this. Could we have two sets of interfaces built up:

`BaseStream` - can we rename this to `Stream`, it makes more sense (we don't use `Base` in other places in yarpc-go I don't think). I rather rename the enum type to something else, perhaps `Streaming`
`ClientStream`
`ServerStream`
`SomethingClientStream`
`SomethingServerStream`

Where `Something` is a name I can't come up with that includes `SendMsg` and `RecvMsg`

File: api/transport/stream.go (hunk 2 above)
Comment (peter-edge) [inline api/transport/stream.go]: Actually I take it back, let's keep it as `Close`

File: transport/x/grpc/stream.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+func (cs *clientStream) Context() context.Context {
+	return cs.ctx
+}
+
+func (cs *clientStream) Request() *transport.Request {
+	return cs.treq
+}
+
+func (cs *clientStream) SendMsg(m io.Reader) error {
```
Comment (peter-edge) [inline transport/x/grpc/stream.go]: I haven't done through this carefully so there's prob a reason but why not just do `[]byte` as the parameter type, or `interface{}` like grpc-go does?

**Comment by peter-edge** [comment]
One other thing - it would be great if the implementation of `Close` on ClientStream could be called multiple times with no effect. I have a `CloseAndRecv` method generated for all client interfaces for client-only streaming, so it would be nice if I had the option to call `Close` if I didn't care about the response object, and if calling `CloseAndRecv` after calling `Close` did the same thing as just calling `CloseAndRecv`

**Comment by peter-edge** [comment]
It would also be great if we have `ClientServerStream` or `BidiStream` or `BidirectionalStream` that was a merge of the interfaces `ClientStream` and `ServerStream` so I could include both of them. I know that right now the merge is just `ClientStream` but this would be weird to assume.

File: transport/x/grpc/handler.go [hunk 5]
```diff
@@ -59,9 +59,87 @@ func (h *handler) handle(srv interface{}, serverStream grpc.ServerStream) error
...
+	if err != nil {
+		return err
+	}
+	switch handlerSpec.Type() {
+	case transport.Unary:
+		return h.handleUnary(ctx, transportRequest, serverStream, streamMethod, start, handlerSpec.Unary())
+	case transport.Stream:
+		return h.handleStream(ctx, transportRequest, serverStream, handlerSpec.Stream())
+	}
+	return yarpcerrors.UnimplementedErrorf("transport grpc does not handle %s handlers", handlerSpec.Type().String())
```
Comment (willhug) [inline transport/x/grpc/handler.go]: This needs to be converted back

File: transport/x/grpc/handler.go [hunk 6]
```diff
@@ -59,9 +59,87 @@ func (h *handler) handle(srv interface{}, serverStream grpc.ServerStream) error
 		return errInvalidGRPCStream
 	}
 
+	streamMethod := stream.Method()
```
Comment (willhug) [inline transport/x/grpc/handler.go]: Need to wrap most of this function in an error converter.

File: api/transport/stream.go (hunk 1 above)
Comment (willhug) [inline api/transport/stream.go]: Talked with @abhinav offline, we'll leave this as is for now, there's not a good way to split this up before we do a 2.0 of yarpc.  So for now streams will have a transport.Request without a "body" field.

One thing we'll just need to do is that for yarpc 2.0 we need to split out the "Metadata" from the transport request and use that here instead.

File: api/transport/clientconfig.go [hunk 7]
```diff
@@ -44,4 +44,9 @@ type ClientConfig interface {
 	// have already been started.
 	GetUnaryOutbound() UnaryOutbound
 	GetOnewayOutbound() OnewayOutbound
+
+	// DO NOT COMMIT INTO MAIN BRANCH.  This is a breaking change against the
+	// YARPC API we can use to play around with streaming.  It should be removed
+	// before we land this code in production.
+	GetStreamOutbound() StreamOutbound
```
Comment (willhug) [inline api/transport/clientconfig.go]: sgtm

File: transport/x/grpc/stream.go (hunk 4 above)
Comment (willhug) [inline transport/x/grpc/stream.go]: I'd be interested in pulling in other opinions on this.  My initial thought was to follow the model we've been using in yarpc requests/responses.  My assumption is that throwing a Reader around any src bytes is pretty low cost which makes it a more flexible api when dealing with different streaming transports.

I'd love to get people's thoughts on whether we should use io.Readers, []bytes, or interface{}  for streaming.

@prashantv @abhinav @akshayjshah @breerly

File: transport/x/grpc/stream.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	msg, err := ioutil.ReadAll(m)
+	if err != nil {
+		return err
+	}
+	return cs.stream.SendMsg(msg)
+}
+
+func (cs *clientStream) RecvMsg() (io.Reader, error) {
```
Comment (willhug) [inline transport/x/grpc/stream.go]: If we make this a ReadCloser, we can Pool the inbound buffers for RecvMsg...

**Comment by peter-edge** [comment]
@willhug I updated the protobuf generation code most of the way, and left todos where it still needs to be finished - I don't want to go any further until the interfaces are fleshed out a bit more.

Additionally, I realized something through building this - you need three types of Handlers in `api/transport`, one for client-only streaming, one for server-only streaming, and one for bidirectional streaming. All will have different APIs that reflect getting a single request (server-only streaming), or returning a single response (client-only streaming). You'll need a `transport.ResponseWriter` for server-only streaming, for example. We don't want to rely on the implementer of each transport doing this properly, it can be abstracted away.

I have this split between streaming types in the protobuf generated code, so we just need it here. We probably either want more splits in the Stream interfaces, or just have one type that covers everything (I'm not including the Stream interfaces in the generated code for now). You'll also need to figure out how to smartly handle the lack of a body on transport.Request in various places.

**Comment by peter-edge** [comment]
Two other things:

- Don't worry about the `eee` stuff right now, I'll remove that when I'm done, I have a directory `eee` that has some testing stuff in it for me to work on the protobuf generated code
- See `encoding/protobuf/protoc-gen-yarpc-go/internal/testing/testing.{pb.yarpc.go,proto}` for an example of the output.

**Comment by peter-edge** [comment]
Also this currently torches my hard-earned almost-80% code coverage on transport/x/grpc right now but whatever lol.

**Comment by willhug** [comment]
We shouldn't increase the API surface any more than we have to.   The bidirectional streaming API we currently have is sufficient for all streaming use cases.  If we want to have a common API that converts a bidirectional stream into a unidirectional one we can make a package to do that.  But it should be the responsibility of the encoding to expose that API since before we get to the encoding we have no information on what kind of stream it is (bidirectional, unidirectional, etc).

File: api/transport/stream.go (hunk 1 above)
Comment (breerly) [inline api/transport/stream.go]: Let's be sure to capture this stuff in the https://github.com/yarpc/yarpc-go/issues/1197 before landing.

File: .codecov.yml [hunk 9]
```diff
@@ -14,6 +14,7 @@ coverage:
 # This list is generated using scripts/generate-cover-ignore.sh
 # All packages with .nocover files in them MUST be listed here
 ignore:
+ - /eee/
```
Comment (peter-edge) [inline .codecov.yml]: Remove before we merge

File: .gitignore [hunk 10]
```diff
@@ -10,6 +10,7 @@
 /vendor
 /TAGS
 
+/eee
```
Comment (peter-edge) [inline .gitignore]: Remove before we merge

File: api/transport/clientconfig.go (hunk 7 above)
Comment (peter-edge) [inline api/transport/clientconfig.go]: Do this

File: encoding/protobuf/inbound.go [hunk 11]
```diff
@@ -107,3 +107,17 @@ func getProtoRequest(ctx context.Context, transportRequest *transport.Request, n
 	}
 	return ctx, call, request, nil
 }
+
+type streamHandler struct {
+	handle     func(ServerStream) error
+}
+
+func newStreamHandler(
```
Comment (peter-edge) [inline encoding/protobuf/inbound.go]: no newline for function

File: encoding/protobuf/outbound.go [hunk 12]
```diff
@@ -125,3 +126,45 @@ func (c *client) buildTransportRequest(ctx context.Context, requestMethodName st
...
+		return nil, err
+	}
+	if transportRequest.Encoding != Encoding && transportRequest.Encoding != JSONEncoding {
+		return nil, yarpcerrors.InternalErrorf("can only use encodings %q or %q, but %q was specified", Encoding, JSONEncoding, transportRequest.Encoding)
+	}
+
+	return c.clientConfig.GetStreamOutbound().CallStream(ctx, transportRequest)
+}
+
+func ToProtoMessage(
```
Comment (peter-edge) [inline encoding/protobuf/outbound.go]: should not be public

File: encoding/protobuf/outbound.go [hunk 13]
```diff
@@ -125,3 +126,45 @@ func (c *client) buildTransportRequest(ctx context.Context, requestMethodName st
...
+	newResponse func() proto.Message,
+) (proto.Message, error) {
+	response := newResponse()
+	if err := unmarshal(encoding, src, response); err != nil {
+		return nil, err
+	}
+	return response, nil
+}
+
+func ToReader(request proto.Message, encoding transport.Encoding) (io.Reader, func(), error){
```
Comment (peter-edge) [inline encoding/protobuf/outbound.go]: should not be public, also neither of these functions belong in this file

File: encoding/protobuf/outbound.go [hunk 14]
```diff
@@ -125,3 +126,45 @@ func (c *client) buildTransportRequest(ctx context.Context, requestMethodName st
...
+	requestData, cleanup, err := marshal(encoding, request)
+	if err != nil {
+		return nil, cleanup, err // TODO wrap error
```
Comment (peter-edge) [inline encoding/protobuf/outbound.go]: no need to wrap error, remove todo

File: encoding/protobuf/outbound.go (hunk 12 above)
Comment (peter-edge) [inline encoding/protobuf/outbound.go]: where is this used?

File: encoding/protobuf/protobuf.go [hunk 15]
```diff
@@ -117,8 +139,16 @@ type Client interface {
...
 		options ...yarpc.CallOption,
 	) (transport.Ack, error)
+	CallStream(
+		ctx context.Context,
+		requestMethodName string,
+		opts ...yarpc.CallOption,
+	) (ClientStream, error)
 }
 
+// TODO replace (maybe)
```
Comment (peter-edge) [inline encoding/protobuf/protobuf.go]: fix this up

File: internal/examples/streaming/generate.sh [hunk 16]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (peter-edge) [inline internal/examples/streaming/generate.sh]: this should all be in `etc/bin/generate.sh`

File: transport/x/grpc/handler.go [hunk 17]
```diff
@@ -59,9 +59,87 @@ func (h *handler) handle(srv interface{}, serverStream grpc.ServerStream) error
...
+	transportRequest, err := h.getBasicTransportRequest(ctx, streamMethod)
+	if err != nil {
+		return err
+	}
+
```
Comment (peter-edge) [inline transport/x/grpc/handler.go]: no newline

File: transport/x/grpc/handler.go [hunk 18]
```diff
@@ -59,9 +59,87 @@ func (h *handler) handle(srv interface{}, serverStream grpc.ServerStream) error
...
+func (h *handler) getBasicTransportRequest(ctx context.Context, streamMethod string) (*transport.Request, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if md == nil || !ok {
+		return nil, yarpcerrors.InternalErrorf("cannot get metadata from ctx: %v", ctx)
+	}
+	transportRequest, err := metadataToTransportRequest(md)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (peter-edge) [inline transport/x/grpc/handler.go]: no newline

File: transport/x/grpc/outbound.go [hunk 19]
```diff
@@ -220,3 +220,59 @@ func invokeErrorToYARPCError(err error, responseMD metadata.MD) error {
...
+
+func (o *Outbound) stream(
+	ctx context.Context,
+	request *transport.Request,
+) (_ transport.ClientStream, retErr error) {
+	md, err := transportRequestToMetadata(request)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (peter-edge) [inline transport/x/grpc/outbound.go]: no newline

File: transport/x/grpc/outbound.go [hunk 20]
```diff
@@ -220,3 +220,59 @@ func invokeErrorToYARPCError(err error, responseMD metadata.MD) error {
...
+	grpcPeer, ok := apiPeer.(*grpcPeer)
+	if !ok {
+		return nil, peer.ErrInvalidPeerConversion{
+			Peer:         apiPeer,
+			ExpectedType: "*grpcPeer",
+		}
+	}
+
```
Comment (peter-edge) [inline transport/x/grpc/outbound.go]: no newline

File: transport/x/grpc/peer.go [hunk 21]
```diff
@@ -159,7 +160,17 @@ func (p *grpcPeer) wait() error {
 	if p.stopped {
 		return p.stoppedErr
 	}
-	p.stoppedErr = <-p.stoppedC
+
+	// TODO Fix in another PR
+	// Stopping the outbound always timed out at the stoppedC in my tests, this
+	// is just a workaround that doesn't solve the core issue.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Comment (peter-edge) [inline transport/x/grpc/peer.go]: ?

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 22]
```diff
@@ -96,7 +157,36 @@ func Build{{$service.GetName}}YARPCProcedures(server {{$service.GetName}}YARPCSe
...
 							Handle: handler.{{$method.GetName}},
-							NewRequest: new{{$service.GetName}}_{{$method.GetName}}YARPCRequest,
+							NewRequest: new{{$service.GetName}}Service{{$method.GetName}}YARPCRequest,
+						},
+					),
+				},
+			{{end}}
+			},
+			StreamHandlerParams: []protobuf.BuildProceduresStreamHandlerParams{
+			{{range $method := clientServerStreamingMethods $service}}{
```
Comment (peter-edge) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: please match the order in the rest of this file - client, server, clientServer

File: encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go [hunk 23]
```diff
@@ -128,6 +218,43 @@ func (c *_{{$service.GetName}}YARPCCaller) {{$method.GetName}}(ctx context.Conte
...
+	return &_{{$service.GetName}}Service{{$method.GetName}}YARPCClient{stream: stream}, nil
+}
+{{end}}
+{{range $method := serverStreamingMethods $service}}
+func (c *_{{$service.GetName}}YARPCCaller) {{$method.GetName}}(ctx context.Context, request *{{$method.RequestType.GoType $packagePath}}, options ...yarpc.CallOption) ({{$service.GetName}}Service{{$method.GetName}}YARPCClient, error) {
+	stream, err := c.client.CallStream(ctx, "{{$method.GetName}}", options...)
+	if err != nil {
+		return nil, err
+	}
+	reader, closer, err := protobuf.ToReader(request, stream.Request().Encoding)
```
Comment (peter-edge) [inline encoding/protobuf/protoc-gen-yarpc-go/internal/lib/lib.go]: oh, now I see where this is

this function needs a better name, and if public, should be in `encoding/protobuf/protobuf.go`

**Comment by peter-edge** [comment]
Please look at the outdated comments too since there's a new push. You also need to add integration testing in `encoding/protobuf/testing`, and clean up the commented out code in examples. Also `transport/x/grpc/integration_test.go`.

**Comment by peter-edge** [comment]
`encoding/protobuf/protoc-gen-yarpc-go/internal/testing/testing.pb.yarpc.go.golden` also needs updating

File: transport/x/grpc/peer.go (hunk 21 above)
Comment (willhug) [inline transport/x/grpc/peer.go]: Yeah, this was why I think there's a race condition in the GRPC peer API, whenever I would stop the grpc outbound the p.stoppedC channel would never close and block indefinitely

File: api/transport/handler.go [hunk 24]
```diff
@@ -65,6 +69,9 @@ func (h HandlerSpec) Unary() UnaryHandler { return h.unaryHandler }
 // Oneway returns the Oneway Handler or nil
 func (h HandlerSpec) Oneway() OnewayHandler { return h.onewayHandler }
 
+// Stream returns the Stream Handler or nil
```
Comment (vivekrsharma) [inline api/transport/handler.go]: nit: Most of the comments in this file or above don end with a period. maybe we want to fix that?

File: transport/x/grpc/handler.go (hunk 5 above)
Comment (peter-edge) [reply transport/x/grpc/handler.go]: There's a bigger overall issue here - streaming isn't handling yarpcerrors. In the code before this PR, everything was done within `handleBeforeErrorConversion`, and a response metadata was used to handle named errors. None of this logic applies to streaming now, so this needs to be weaved into streaming error handling.

Comment (peter-edge) [reply transport/x/grpc/handler.go]: We also need to handle opentracing before this can be merged.

**Comment by peter-edge** [comment]
@willhug I did some cleanups to make this easier, hope this helps

**Comment by peter-edge** [comment]
Tests passing again :-)

**Comment by willhug** [comment]
Closing in favor of the new streaming prs

**Pull request closed without merging**